pkg/service/admission/endpoint_admission.go

```diff
@@ -0,0 +1,64 @@
+package admission
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/openshift/origin/pkg/cmd/server/bootstrappolicy"
+
+	kadmission "k8s.io/kubernetes/pkg/admission"
+	kapi "k8s.io/kubernetes/pkg/api"
+	apierrs "k8s.io/kubernetes/pkg/api/errors"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+)
+
+const ExplicitEndpointPluginName = "ExplicitEndpointRestriction"
+
+func init() {
+	kadmission.RegisterPlugin("ExplicitEndpointRestriction", func(client clientset.Interface, config io.Reader) (kadmission.Interface, error) {
+		return NewExplicitEndpointRestriction(client), nil
+	})
+}
+
+type explicitEndpointRestriction struct {
+	*kadmission.Handler
+
+	kclient clientset.Interface
+}
+
+var _ kadmission.Interface = &explicitEndpointRestriction{}
+
+// NewExplicitEndpointRestriction creates a new endpoint restriction admission plugin.
+func NewExplicitEndpointRestriction(client clientset.Interface) *explicitEndpointRestriction {
+	return &explicitEndpointRestriction{
+		Handler: kadmission.NewHandler(kadmission.Create, kadmission.Update),
+		kclient: client,
+	}
+}
+
+// Admit determines if the endpoint should be admitted based on its service
+func (r *explicitEndpointRestriction) Admit(a kadmission.Attributes) error {
+	if a.GetResource().GroupResource() != kapi.Resource("endpoints") {
+		return nil
+	}
+	ep, ok := a.GetObject().(*kapi.Endpoints)
+	if !ok {
+		return nil
+	}
+
+	// The Endpoint controller is always allowed to add Endpoints
+	if a.GetUserInfo().GetName() == bootstrappolicy.MasterUsername {
+		return nil
+	}
+
+	svc, err := r.kclient.Core().Services(a.GetNamespace()).Get(ep.Name)
+	if err != nil {
+		return apierrs.NewNotFound(a.GetResource().GroupResource(), ep.Name)
+	}
+
+	if svc.Spec.Selector != nil {
+		return apierrs.NewForbidden(a.GetResource().GroupResource(), ep.Name, fmt.Errorf("can't add explicit Endpoints to Service with a selector"))
+	}
+
+	return nil
```

Comment:
Yeah, that's the safer and more compact change - doesn't require us to
solve races.


On Jun 16, 2016, at 12:45 PM, Jordan Liggitt <notifications@github.com>
wrote:

In pkg/service/admission/endpoint_admission.go
<https://github.com/openshift/origin/pull/9383#discussion_r67381183>:

> +
> +	// The Endpoint controller is always allowed to add Endpoints
> +	if a.GetUserInfo().GetName() == bootstrappolicy.MasterUsername {
> +		return nil
> +	}
> +
> +	svc, err := r.kclient.Core().Services(a.GetNamespace()).Get(ep.Name)
> +	if err != nil {
> +		return apierrs.NewNotFound(a.GetResource().GroupResource(), ep.Name)
> +	}
> +
> +	if svc.Spec.Selector != nil {
> +		return apierrs.NewForbidden(a.GetResource().GroupResource(), ep.Name, fmt.Errorf("can't add explicit Endpoints to Service with a selector"))
> +	}
> +
> +	return nil

I wasn't expecting a user to be able to point to any IP they want just
because the service doesn't have a selector. I was expecting something to
prevent regular users from setting endpoint IPs within the service or pod
CIDRs

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/9383/files/b21ea8af78d57db74065dacec6a4bc8f93496912#r67381183>,
or mute the thread
<https://github.com/notifications/unsubscribe/ABG_p3vGWwVyHjYYmBgNTWvd3SzrrzHTks5qMX20gaJpZM4I3hI->
.